File: src/encoder/protocolencoder.cpp
#include "encoder/protocolencoder.h"

#include "flagdata/dataheader.h"
#include "flagdata/datacontent.h"
#include "flagdata/datasize.h"
#include "flagdata/dataverify.h"
#include "flagdata/dataend.h"

PROTOCOL_CODEC_NAMESPACE_BEGIN

void ProtocolEncoder::addType(int type, ProtocolTypeCodecInterface* codec) {
    typeEncoders.insert(type, codec);
}

ProtocolEncoder::~ProtocolEncoder() {
    qDeleteAll(typeEncoders);
}

ProtocolTypeCodecInterface *ProtocolEncoder::getCodec(int type) {
    return typeEncoders[type];
}

QByteArray ProtocolEncoder::encodeFrame(const QByteArray &content, int dataType) {
    QByteArray buffer;
    int contentOffset = 0;
    printInfo([&] {
        return QString("encode type %1 begin, content size: %2").arg(dataType).arg(content.size());
    });
    for (const auto& flag : protocolFlags) {
        switch (flag->flag) {
            case ProtocolFlag::Flag_Header: {
                auto headerFlag = qSharedPointerCast<ProtocolFlagDataHeader>(flag);
                buffer.append(headerFlag->target);
            }
                break;
            case ProtocolFlag::Flag_Size: {
                auto sizeFlag = qSharedPointerCast<ProtocolFlagDataSize>(flag);
                int contentSize = content.size() + mTypeByteSize;
                auto sizeBuff = (char*) malloc(sizeFlag->byteSize);
                for (int i = 0; i < sizeFlag->byteSize; i++) {
                    sizeBuff[i] = (char)((contentSize >> (i * 8)) & 0xff);
                }
                buffer.append(sizeBuff, sizeFlag->byteSize);
                sizeFlag->dataSize = contentSize;
                free(sizeBuff);
            }
                break;
            case ProtocolFlag::Flag_Content: {
                contentOffset = buffer.size();
                buffer.append((char*)&dataType, mTypeByteSize).append(content);
            }
                break;
            case ProtocolFlag::Flag_Verify: {
                auto verifyFlag = qSharedPointerCast<ProtocolFlagDataVerify>(flag);
                auto verifyCode = verifyFlag->getVerifyCode(buffer, contentOffset);
                printInfo([&] {
                    return "content verify code: " + verifyCode.toHex(' ');
                });
                buffer.append(verifyCode);
            }
                break;
            case ProtocolFlag::Flag_End: {
                auto endFlag = qSharedPointerCast<ProtocolFlagDataEnd>(flag);
                buffer.append(endFlag->target);
            }
                break;
            default:
                break;
        }
    }
    return buffer;
}


PROTOCOL_CODEC_NAMESPACE_END

File: src/main/kotlin/codec/ProtocolEncoder.kt
package com.daonvshu.protocol.codec.codec

import com.daonvshu.protocol.codec.ProtocolFlag.*
import com.daonvshu.protocol.codec.flagdata.ProtocolFlagDataEnd
import com.daonvshu.protocol.codec.flagdata.ProtocolFlagDataHeader
import com.daonvshu.protocol.codec.flagdata.ProtocolFlagDataSize
import com.daonvshu.protocol.codec.flagdata.ProtocolFlagDataVerify
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

class ProtocolEncoder : ProtocolCodec() {

    val typeEncoders = mutableMapOf<Int, (Any) -> ByteArray>()

    fun encodeFrame(data: ByteArray, dataType: Int): ByteArray {
        val buffer = ByteArrayOutputStream()
        var contentOffset = 0

        for (flag in flags) {
            when (flag.flagType) {
                FlagHeader -> {
                    val headerFlag = flag as ProtocolFlagDataHeader
                    buffer.write(headerFlag.target)
                }

                FlagSize -> {
                    val sizeFlag = flag as ProtocolFlagDataSize
                    val contentSize = data.size + typeByteSize

                    // 低位优先字节序写入contentSize
                    val sizeBuff = ByteArray(sizeFlag.byteSize)
                    for (i in 0 until sizeFlag.byteSize) {
                        sizeBuff[i] = ((contentSize shr (i * 8)) and 0xFF).toByte()
                    }
                    buffer.write(sizeBuff)
                    sizeFlag.dataSize = contentSize
                }

                FlagContent -> {
                    contentOffset = buffer.size()
                    // 写入dataType，低位优先（小端）
                    val typeBytes = ByteBuffer.allocate(typeByteSize)
                        .order(ByteOrder.LITTLE_ENDIAN)
                        .putInt(dataType)
                        .array()
                    buffer.write(typeBytes)
                    if (data.isNotEmpty()) {
                        buffer.write(data)
                    }
                }

                FlagVerify -> {
                    val verifyFlag = flag as ProtocolFlagDataVerify
                    val currentBuffer = buffer.toByteArray()
                    val verifyCode = verifyFlag.getVerifyCode(currentBuffer, contentOffset)
                    buffer.write(verifyCode)
                }

                FlagEnd -> {
                    val endFlag = flag as ProtocolFlagDataEnd
                    buffer.write(endFlag.target)
                }

                FlagMax -> {}
            }
        }

        return buffer.toByteArray()
    }
}